demo_modules/wind/vector_field.js

```diff
@@ -0,0 +1,258 @@
+		(pφ[0] - x) / hφ,
+		(pφ[1] - y) / hφ
+	];
+}
+
+/**
+ * Calculate distortion of the vector caused by the shape of the projection at point (x, y). The
+ * vector is modified in place and returned by this function.
+ */
+function distort(projection, λ, φ, x, y, scale, vector) {
+	var u = vector[0] * scale;
+	var v = vector[1] * scale;
+	var d = distortion(projection, λ, φ, x, y);
+
+	// Scale distortion vectors by u and v, then add.
+	vector[0] = d[0] * u + d[2] * v;
+	vector[1] = d[1] * u + d[3] * v;
+	return vector;
+}
+
+class Mask {
+  constructor(projection) {
+		// TODO(bmt): Get these from wall geometry.
+		this.width = 1920;
+		this.height = 1080;
+
+		// Create a detached canvas, draw an opaque sphere that represents visible
+		// points.
+		var canvas = d3.select(document.createElement("canvas"))
+			.attr("width", this.width).attr("height", this.height).node();
+    const context = canvas.getContext('2d');
+
+    const projectedPath = d3.geo.path().projection(projection).context(context);
+
+		const mask = projectedPath({type: "Sphere"});
+		context.fillStyle = "rgba(255, 0, 0, 1)";
+		context.fill();
+
+		// layout: [r, g, b, a, r, g, b, a, ...]
+		this.imageData = context.getImageData(0, 0, this.width, this.height).data;
+  }
+
+	isVisible(x, y) {
+		var i = (y * this.width + x) * 4;
+		return this.imageData[i + 3] > 0;  // non-zero alpha means pixel is visible
+	}
+
+	set(x, y, rgba) {
+		var i = (y * this.width + x) * 4;
+		this.imageData[i] = rgba[0];
+		this.imageData[i + 1] = rgba[1];
+		this.imageData[i + 2] = rgba[2];
+		this.imageData[i + 3] = rgba[3];
+		return this;
+	}
+}
+
+class VectorField {
+  constructor(columns, bounds, mask) {
+    this.columns = columns;
+    this.bounds = bounds;
+    this.overlay = new ImageData(mask.imageData, 1920, 1080);
+  }
+
+	/**
+	 * @returns {Array} wind vector [u, v, magnitude] at the point (x, y), or [NaN, NaN, null] if wind
+	 *          is undefined at that point.
+	 */
+	vector(x, y) {
+		var column = this.columns[Math.round(x)];
+		return column && column[Math.round(y)] || NULL_VECTOR;
+	}
+
+	/**
+	 * @returns {boolean} true if the field is valid at the point (x, y)
+	 */
+	isDefined(x, y) {
+		return this.vector(x, y)[2] !== null;
+	}
+
+	/**
+	 * @returns {boolean} true if the point (x, y) lies inside the outer boundary of the vector field, even if
+	 *          the vector field has a hole (is undefined) at that point, such as at an island in a field of
+	 *          ocean currents.
+	 */
+	isInsideBoundary(x, y) {
+		return this.vector(x, y) !== NULL_VECTOR;
+	}
+
+	// Frees the massive "columns" array for GC. Without this, the array is leaked (in Chrome) each time a new
+	// field is interpolated because the field closure's context is leaked, for reasons that defy explanation.
+  // TODO(bmt): Make sure this is actually necessary (not my comment).
+	release() {
+    this.columns = [];
+	}
+
+	// TODO(bmt): Figure out what's happening here (not my comment).
```

Comment:
b/c it tries a guess, then retries. It would be nice if there were some way to avoid the loop.